File: highlevel/highlevel_controllers/src/indefinite_nav.cpp
/*********************************************************************
 * Software License Agreement (BSD License)
 * 
 *  Copyright (c) 2008, Willow Garage, Inc.
 *  All rights reserved.
 * 
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 * 
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of the Willow Garage nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 * 
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/


#include <IndefiniteNav.hh>
#include <math.h>

using namespace std;


namespace ros {
  namespace highlevel_controllers {

    IndefiniteNav::IndefiniteNav() : ros::Node("indefinite_nav_node")
    {
      addGoal (24.6, 30, 0.66);
      addGoal (17.5, 31.3, -1.02);
      addGoal (30.45, 37.6, 1.11);
      addGoal (33.75, 31.4, -1.53);
      addGoal (28.85, 24.5, -0.63);
      addGoal (25.7, 22.5, -1.74);
      addGoal (17.6, 14.2, 1.78);
      addGoal (33.1, 12.1, 1.64);
      first = true;
    }



    void IndefiniteNav::run() 
    {
      list<Planner2DGoal>::iterator i;

      cout << "In IndefiniteNav.run\n";
      advertise<Planner2DGoal>("goal", 10);
      subscribe<Planner2DState>("state", state_msg, &IndefiniteNav::stateCallback, 10);
	
      sleep(4);
      while (true) {

	for (i=msgs.begin(); i!=msgs.end(); i++) {
	  
	  if (first || fabs(state_msg.pos.x - state_msg.goal.x) + fabs(state_msg.pos.y - state_msg.goal.y) < 1) {
	    printf ("Dispatching goal %f, %f, %f", i->goal.x, i->goal.y, i->goal.th);
	    publish ("goal", *i);
	    sleep (2); // Should avoid need for lock
	    first = false;
	  }
	}
      }
    }





    void IndefiniteNav::addGoal (double x, double y, double th) {
	
      Planner2DGoal g;
      g.enable = 1;
      g.goal.x = x;
      g.goal.y = y;
      g.goal.th = th;
      msgs.push_back(g);

    }



    void IndefiniteNav::stateCallback ()
    {
    }



  }
}

















int main (int argc, char** argv)
{
  ros::init(argc,argv);
  ros::highlevel_controllers::IndefiniteNav node;
  node.run();
  ros::fini();
  return(0);
}
  

File: highlevel/executive_python/src/indefinite_nav.py
#! /usr/bin/python
#***********************************************************
#* Software License Agreement (BSD License)
#*
#*  Copyright (c) 2008, Willow Garage, Inc.
#*  All rights reserved.
#*
#*  Redistribution and use in source and binary forms, with or without
#*  modification, are permitted provided that the following conditions
#*  are met:
#*
#*   * Redistributions of source code must retain the above copyright
#*     notice, this list of conditions and the following disclaimer.
#*   * Redistributions in binary form must reproduce the above
#*     copyright notice, this list of conditions and the following
#*     disclaimer in the documentation and/or other materials provided
#*     with the distribution.
#*   * Neither the name of the Willow Garage nor the names of its
#*     contributors may be used to endorse or promote products derived
#*     from this software without specific prior written permission.
#*
#*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
#*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
#*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
#*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
#*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
#*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
#*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
#*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
#*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#*  POSSIBILITY OF SUCH DAMAGE.
#* 
#* Author: Eitan Marder-Eppstein
#***********************************************************
import roslib
roslib.load_manifest('executive_python')
import rospy
import random
from std_msgs.msg import *
from robot_msgs.msg import *
from battery_monitor_adapter import *
from recharge_adapter import *
from navigation_adapter import *

class IndefiniteNav:
  def __init__(self, goals, navigator, cycle_time):
    rospy.init_node("IndefiniteNav", anonymous=True)
    self.goals = goals
    self.navigator = navigator
    self.cycle_time = cycle_time
    self.state = "nav"
    self.current_goal = self.goals[0]

  def legalStates(self):
    return self.navigator.legalState()

  def doCycle(self):
    #make sure that all adapters have legal states
    if self.legalStates():
      if self.state == "nav":
        """ 
          State Transitions from nav:
            1) nav --- robot is inactive initially, reaches a goal, or timeout is reached ---> nav
            2) nav --- robot is inactive and should be pursuing the current goal (move_base crashed) ---> nav
        """
        if self.navigator.goalReached() or (not self.navigator.active() and self.current_goal == None) or self.navigator.timeUp():
          self.current_goal = self.goals[random.randint(0, len(self.goals) - 1)]
          self.navigator.sendGoal(self.current_goal, "map")
          print "nav --> nav"
        elif not self.navigator.active() and self.current_goal != None:
          self.navigator.sendGoal(self.current_goal, "map")
          print "nav --> nav"
    else:
      if not self.navigator.legalState():
        print("Waiting on %s to be published" % (self.navigator.state_topic))
        rospy.logout("Waiting on %s to be published" % (self.navigator.state_topic))

  def run(self):
    while not rospy.is_shutdown():
      start = rospy.get_time()
      self.doCycle()
      end = rospy.get_time()
      sleep_time = self.cycle_time - (end - start)
      if sleep_time > 0:
        rospy.sleep(sleep_time)
      else:
        print("IndefiniteNav missed cycle time of %.2f seconds" % (self.cycle_time))
        rospy.logwarn("IndefiniteNav missed cycle time of %.2f seconds" % (self.cycle_time))

if __name__ == '__main__':
  try:
    #navigator = NavigationAdapter(30, 300, "/move_base_node/feedback", "/move_base_node/activate")
    navigator = NavigationAdapter(30, 300, "/move_base_node/feedback", "/goal")

    goals = [
     [[18.912, 28.568, 0.000], [0.000, 0.000, 0.000, 1.000]],
     [[12.118, 39.812, 0.000], [0.000, 0.000, 0.000, 1.000]],
     [[47.350, 40.655, 0.000], [0.000, 0.000, -0.707, 0.707]],
     [[54.162, 20.087, 0.000], [0.000, 0.000, -0.160, 0.987]],
     [[50.659, 6.916, 0.000], [0.000, 0.000, 0.924, -0.383]],
     [[18.542, 12.301, 0.000], [0.000, 0.000, 0.906, -0.424]]
     ]


    indefinite_nav = IndefiniteNav(goals, navigator, 1.0)
    indefinite_nav.run()
  except KeyboardInterrupt, e:
    pass
  print "exiting"


